### PR TITLE
Add guide for new io formats

### PR DESCRIPTION
**Pull request by felixdivo**
This adds a guide on how to add new file format support.

Added in response to #506.

### CODE REVIEW

**Comment by danielhrisca** [comment]
@felixdivo 

each Writer is meant to handle a single bus channel? For example if we have a vector HW and we use bus channels 0 and 1 then the used should instantiate two Write objects for each of them?

**Comment by felixdivo** [comment]
Writers are just dumb `can.Listeners`, which receive messages in the `on_message_received()` method. The writer does not care about the bus that this message came from or how it was brought there, except maybe that the `Message.channel` attribute is set that then gets serialized. So there may be zero to many  writers serializing the messages of zero to many buses. Distributing is often done by using a [Notifier](https://python-can.readthedocs.io/en/stable/api.html#notifier). Does that answer you question?

**Comment by danielhrisca** [comment]
I got the point. I'll think about a strategy to handle this

File: doc/internal-api.rst [hunk 1]
```diff
@@ -66,19 +67,52 @@ methods:
...
+
+
+Adding support for new file formats
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This assumes that you want to add a new file format, called *canstore*.
+
+1. Create a new module: *can/io/canstore.py*
+   (*or* simply copy some existing one like *can/io/csv.py*)
+2. Implement a reader ``CanstoreReader`` (which extends :class:`can.io.generic.BaseIOHandler`).
```
Comment (hardbyte) [inline doc/internal-api.rst]: A `Reader` doesn't *have* to extend `BaseIOHandler`, that is a good starting point for any *file* based format though.

A small suggestion is to add a `~` to the path to improve readability in the rendered docs (but ensuring the link works).

```suggestion
2. Implement a reader ``CanstoreReader`` (which extends :class:`~can.io.generic.BaseIOHandler`).
```

File: doc/internal-api.rst [hunk 2]
```diff
@@ -66,19 +67,52 @@ methods:
...
+The reader usually extends :class:`can.io.generic.BaseIOHandler`, while
+the writer additionally extends :class:`can.Listener`,
+to be able to be passed directly to a :class:`can.Notifier`.
+
+
+
+Adding support for new file formats
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This assumes that you want to add a new file format, called *canstore*.
```
Comment (hardbyte) [inline doc/internal-api.rst]:
```suggestion
This assumes that you want to add a new file format, called *canstore*. Ideally add both reading and writing support for the new file format, although this is not strictly required.
```

File: doc/internal-api.rst [hunk 3]
```diff
@@ -66,19 +67,52 @@ methods:
...
+
+1. Create a new module: *can/io/canstore.py*
+   (*or* simply copy some existing one like *can/io/csv.py*)
+2. Implement a reader ``CanstoreReader`` (which extends :class:`can.io.generic.BaseIOHandler`).
+   Besides from a constructor, only ``__iter__(self)`` needs to be implemented.
+3. Implement a writer ``CanstoreWriter`` (which extends :class:`can.io.generic.BaseIOHandler` and :class:`can.Listener`).
```
Comment (hardbyte) [inline doc/internal-api.rst]: Again they are optional base classes. The only requirement is the writer must have a `on_message_received` method.

File: doc/internal-api.rst (hunk 1 above)
Comment (felixdivo) [reply doc/internal-api.rst]: I thought that it might be nice to know where this class sits, since this is documentation for the library developers.

**Pull request merged**